### ISSUE TITLE
Model loading and different results on multiple evaluating of the model

### ISSUE THREAD

**Issue by DiplEng** [keras-team/keras#7799]
Hello everyone,

I've just used for the first time the ModelCheckpoint function to save the best model (best_model = True) and wanted to test its performance. When the model was saved it said that the "val_acc" was at 83.3% before saving. I loaded the model and used the evaluate_generator on validation_generator but the result for "val_acc" was 0.639. I got confused and used it again and got 0.654 and then 0.647, 0.744 and so on. Questions are:

1. Am I loading the model correctly, and if not what did I miss? Why is the result so much different?
2. Why are the results between different evaluate_generator executions different?

important part of the code (ResNet50 fine-tuning):

```
model.compile(loss='categorical_crossentropy',
              optimizer=optimizers.SGD(lr=1e-4, momentum=0.9),
              metrics=['accuracy'])
checkpointer = ModelCheckpoint(filepath='/tmp/weights.hdf5', monitor = "val_acc", verbose=1, save_best_only=True)
# prepare data augmentation configuration
train_datagen = ImageDataGenerator(
    rescale = 1./ 255,
    shear_range = 0.2,
    zoom_range = 0.2,
    horizontal_flip = True)
test_datagen = ImageDataGenerator(rescale=1. / 255)
train_generator = train_datagen.flow_from_directory(
    train_data_dir,
    target_size = (img_height, img_width),
    batch_size = batch_size)
validation_generator = test_datagen.flow_from_directory(
    validation_data_dir,
    target_size = (img_height, img_width),
    batch_size = batch_size)
# fine-tune the model
model.fit_generator(
    train_generator,
    steps_per_epoch = math.ceil(train_samples/batch_size),
    epochs=100,
    workers = 120,
    validation_data=validation_generator,
    validation_steps=math.ceil(val_samples/batch_size),
    callbacks=[checkpointer])
model.load_weights(filepath='/tmp/weights.hdf5')
model.predict_generator(validation_generator, steps = math.ceil(val_samples/batch_size) )
temp_model = load_model('/tmp/weights.hdf5')
temp_model.evaluate_generator(validation_generator, steps = math.ceil(val_samples/batch_size), workers = 120)
```

**Comment by grafael** [comment]
The problem is that you are using **random seeds** . 
You need to set some constant seeds like:
```python
from numpy.random import seed
from tensorflow import set_random_seed

seed(1337)
set_random_seed(1337)
```

You can get more information about reproducibility [here](https://machinelearningmastery.com/reproducible-results-neural-networks-keras/)

**Comment by DiplEng** [comment]
Hi, 
Thanks for the answer. I am aware of these problems, but my problem lies elsewhere (here Jason described why do model result between different trainings). With me, I train a model, and use the predict function **a few times in a row** (without any training in between) and every single time I get different results. Also one would expect that the loaded model shows the same results as when saved once loaded. The validation generator has the shuffle parameter on, but I use all the data, so it shouldn't be a problem. 
Thanks!

**Comment by bdwyer2** [comment]
Are you sure `val_samples` is set to the correct value?

**Comment by DiplEng** [comment]
Ok, the problem was the following - BATCH_SIZE!!! It took me ages to figure this one out - 

    steps = math.ceil(val_samples/batch_size)

Due to the fact that the batch_size was not a divisor of number_of_samples it created problems. Some small errors occured also from setting `workers` variable -  using GPU it makes no sense to actually use it.  Once I used a real divisor of the val_samples it worked like a charm and reproducible!
